__tests__/browser.test.js

```diff
@@ -0,0 +1,275 @@
+const urlEncodedWithAuthHar = require('./__fixtures__/urlencoded-with-auth.har.json');
+
+beforeAll(async () => {
+  await page.exposeFunction('har_full', () => harExamples.full);
+  await page.exposeFunction('har_jsonWithAuthHar', () => jsonWithAuthHar);
+  await page.exposeFunction('har_multipartData', () => harExamples['multipart-data']);
+  await page.exposeFunction('har_multipartFile', () => harExamples['multipart-file']);
+  await page.exposeFunction('har_multipartFormData', () => harExamples['multipart-form-data']);
+  await page.exposeFunction('har_textPlain', () => harExamples['text-plain']);
+  await page.exposeFunction('har_urlEncodedWithAuthHar', () => urlEncodedWithAuthHar);
+});
+
+beforeEach(async () => {
+  await page.goto(SERVER_URL, { waitUntil: 'load' });
+
+  await page.addScriptTag({
+    content: `const SERVER_URL = "${SERVER_URL}";`,
+  });
+
+  await page.evaluate(() => {
+    /**
+     * Take a given HAR object and rewrite its URL to our debug server in order to test requests against it.
+     *
+     * @param {Object} har
+     * @returns {Object}
+     */
+    window.debugHar = har => {
+      // eslint-disable-next-line no-param-reassign
+      har.log.entries[0].request.url = `${SERVER_URL}/debug`;
+      return har;
+    };
+
+    /**
+     * Take a `Request` object and manually serialize it down into an object of the data we're looking for.
+     *
+     * Why? Since `constructRequest` returns a `Request` object, which is not serializable, for our tests we need to
+     * have Puppeteer return an object containing the properties and data we're looking for in order for us to be able
+     * to run assertions within Jest.
+     *
+     * Why not use `page.exposeFunction`? That method unfortunately serializes all parameters that are supplied to
+     * callbacks created with it, and as `Request` can't be serialized we'll end up with an empty object. Also since
+     * we don't need to pass data from the Node layer into Puppetter, as you normally would with `exposeFunction`, we
+     * can skirt that and simply create what we need on the page instance.
+     *
+     * @link https://github.com/puppeteer/puppeteer/issues/1750
+     * @link https://pptr.dev/#?product=Puppeteer&version=v5.2.1&show=api-pageevaluatepagefunction-args
+     * @param {Request} req
+     * @returns {Object}
+     */
+    window.serializeRequest = async req => {
+      return {
+        url: req.url,
+        method: req.method,
+        headers: Object.fromEntries(req.headers.entries()),
+        credentials: req.credentials,
+        cookies: document.cookie,
+        body: (await req.text()).replace(/\r\n/g, '\n'),
+      };
+    };
+  });
+});
+
+describe('#fetch', () => {
+  it('should not make a request with a custom user agent if specified', async () => {
+    const req = await page.evaluate(async () => {
+      const har = debugHar(await window.har_full());
+      return fetchHar(har, 'test-app/1.0').then(res => res.json());
```

Comment:
So does this run within the context of the puppeteer instance? How does fetchHar get exposed to the browser? 

🤯

Reply:
Yeah this runs in the browser. `fetchHar` gets exposed by way of this [server that jest-puppeteer spins up](https://github.com/readmeio/fetch-har/blob/master/__tests__/server.js). That runs an inline instance of Webpack, compiles the library into `index.min.js`, exported as a global `fetchHar` variable ([see here](https://github.com/readmeio/fetch-har/blob/master/__tests__/server.js#L15)), and then loads that minified JS in the browser.

I cringe at this flow potentially busting with a Webpack upgrade, but hey it works pretty well right now!